docs: coordinated workers doc (#295)
canonical/observability-stack: docs/reference/coordinated-workers-meta-roles.md

---
myst:
 html_meta:
  description: "Reference for the worker roles and meta-roles available in the Mimir, Loki, and Tempo coordinated worker deployments in COS."
---

# Coordinated worker roles and meta-roles
Worker meta-roles and roles for Mimir, Loki, and Tempo in COS HA. See the [coordinated workers explanation guide](/explanation/architecture/coordinated-workers.md) for an overview of the coordinator-worker pattern in COS.

## Mimir

### Meta-roles

| Meta-role | Constituent roles |
|---|---|
| `read` | `querier`, `query-frontend` |
| `write` | `distributor`, `ingester` |
| `backend` | `alertmanager`, `compactor`, `overrides-exporter`, `query-scheduler`, `ruler`, `store-gateway` |
| `all` | `compactor`, `distributor`, `ingester`, `querier`, `query-frontend`, `ruler`, `store-gateway` |

### Roles

> | Role | Part of meta-role | Min. deployment | Recommended units |

Reviewer: Should we rename this header to `Required`? It is not clear to me that "minimal deployment": `yes` refers to -> it is required to required / not required. This gives the impression that I can deploy a worker with any of these rows set as `min. deployment: True` and I will have a working `minimal deployment`, which is not true.
```suggestion
| Role | Part of meta-role | Required | Recommended units |
```

---
myst:
 html_meta:

>   description: "Reference for the worker roles and meta-roles available in the Mimir, Loki, and Tempo coordinated worker deployments in COS HA."

Reviewer 2: ```suggestion
  description: "Reference for the worker roles and meta-roles available in the Mimir, Loki, and Tempo coordinated worker deployments in COS."
```

---

> # Coordinated worker roles and meta-roles

Reviewer 2: We should probably emphasize that roles and meta roles are grafana's concepts, not ours.

> In COS HA, each telemetry backend (Mimir, Loki, Tempo) is deployed using the [coordinator-worker pattern](../explanation/architecture/coordinated-workers.md). Workers are configured with one or more *roles*, which determine which internal services the workload process will run.

Reviewer 3: I like that you've separated out the reference material, but there's too much explanation in it.

I'd recommend you move any necessary intro details here into the explanation, and have a super short 1-2 sentences intro at the top here, something basic like "Worker meta-roles and roles for Mimir, Loki, and Tempo in COS HA. See {explanation doc link] for an overview of the coordinator-worker pattern in COS."

Reviewer 3: The main thing in your doc is the meta-role / role names. If you don't think that would fit in your explanation, it's also fine to keep it here, but still I'd keep them super short, more like "This reference lists worker meta-roles and roles for Mimir, Loki, and Tempo in COS HA. Meta-roles are XYZ, and roles are XYZ, See........" (or put the meta roles / roles in bulletpoints)

> In COS HA, each telemetry backend (Mimir, Loki, Tempo) is deployed using the [coordinator-worker pattern](../explanation/architecture/coordinated-workers.md). Workers are configured with one or more *roles*, which determine which internal services the workload process will run.

Reviewer 2: ```suggestion
In COS, each telemetry backend (Mimir, Loki, Tempo) is deployed using the [coordinator-worker pattern](../explanation/architecture/coordinated-workers.md). Workers are configured with one or more *roles*, which determine which internal services the workload process will run.
```

> *Meta-roles* are named shortcuts that expand to a predefined set of roles. They allow operators to deploy common groupings of roles without having to enumerate each one individually. For example, assigning `role-write=true` to a Mimir worker is equivalent to enabling both the `distributor` and `ingester` roles.

Reviewer 2: ```suggestion
*Meta-roles* are named shortcuts that expand to a predefined set of roles. They allow operators to deploy common groupings of roles without having to enumerate each one individually. For example, setting the charm config option `role-write=true` to a Mimir worker is equivalent to enabling both the `distributor` and `ingester` roles.
```

A deployment is considered:
- **Consistent** (or coherent) when all roles in the *minimal deployment* set are covered by at least one worker unit.

> - **Recommended** when all roles in the *recommended deployment* set are covered with the suggested number of units.

Reviewer 2: This is explanation. You already have a good table below. I think we can drop this, or move it next to the table.

Author: Sounds good, I've dropped the table.
